real_preprocessing/src/opt_visualization.cpp

```diff
@@ -0,0 +1,220 @@
+	pose_msg.orientation.y = quat.y();
+	pose_msg.orientation.z = quat.z();
+	pose_msg.orientation.w = quat.w();
+
+	return pose_msg;
+}
+
+//populates a tag marker message with a world_T_tag pose of the indexed tag in the targets.yaml file 
+void loadTagMarker(int tag_index, visualization_msgs::Marker& marker, geometry_msgs::Pose& pose_msg)
+{
+	YAML::Node YAML_handle= YAML::LoadFile(getTargetPath());
+	pose_msg = loadMarkerPose(tag_index);
+
+	marker.header.frame_id = "world"; 
+	marker.header.stamp = ros::Time();
+	marker.ns = "tag_" + std::to_string(YAML_handle["targets"][tag_index]["targetID"].as<int>());
+	marker.id = YAML_handle["targets"][tag_index]["targetID"].as<int>();
+	marker.type = visualization_msgs::Marker::CUBE;
+	marker.action = visualization_msgs::Marker::ADD;
+	marker.pose = pose_msg;
+	marker.scale.x = YAML_handle["targets"][tag_index]["obj_points_in_target"]["2"][0].as<double>()*2;
+	marker.scale.y = YAML_handle["targets"][tag_index]["obj_points_in_target"]["2"][1].as<double>()*2;
+	marker.scale.z = 0.0005;
+	marker.color.a = 1.0;
+	marker.color.r = 1;
+	marker.color.g = 1;
+	marker.color.b = 1;
+	marker.lifetime = ros::Duration();
+}
+
+//generates the trajectory marker with all past camera poses
+visualization_msgs::Marker trajMarker(geometry_msgs::PoseArray campose_array)
+{
+	visualization_msgs::Marker traj_marker;
+	traj_marker.header.frame_id = "world"; 
+	traj_marker.header.stamp = ros::Time();
+	traj_marker.ns = "camera_trajectory";
+	traj_marker.id = -2;
+	traj_marker.type = visualization_msgs::Marker::LINE_STRIP;
+	traj_marker.action = visualization_msgs::Marker::ADD;
+	traj_marker.scale.x = 0.002;
+	traj_marker.color.a = 1.0;
+	traj_marker.color.r = 1;
+	traj_marker.color.g = 1;
+	traj_marker.color.b = 0;
+	traj_marker.colors.push_back(traj_marker.color);
+	traj_marker.lifetime = ros::Duration();
+
+	traj_marker.points.clear();
+	for (int i = 0; i != campose_array.poses.size(); i++)
+	{
+		traj_marker.points.push_back(campose_array.poses[i].position);
+	}
+	return traj_marker;
+}
+
+geometry_msgs::Pose loadCamPose(int filenum)
+{
+	YAML::Node YAML_handle = YAML::LoadFile(getDetfilePath(filenum));
+
+	geometry_msgs::Pose cam_pose;
+	cam_pose.position.x = YAML_handle["world_T_camera"]["translation"][0].as<double>();
+	cam_pose.position.y = YAML_handle["world_T_camera"]["translation"][1].as<double>();
+	cam_pose.position.z = YAML_handle["world_T_camera"]["translation"][2].as<double>();
+
+	double rod_vec[3] = {YAML_handle["world_T_camera"]["rotation"][0].as<double>(),
+							YAML_handle["world_T_camera"]["rotation"][1].as<double>(),
+							YAML_handle["world_T_camera"]["rotation"][2].as<double>()};
+
+	Eigen::Quaterniond quat = rodrigToQuat(rod_vec);
+	cam_pose.orientation.x = quat.x();
+	cam_pose.orientation.y = quat.y();
+	cam_pose.orientation.z = quat.z();
+	cam_pose.orientation.w = quat.w();
+
+	return cam_pose;
+}
+
+//generates camera marker with world_T_cam pose from given file number 
```

Comment:
Populate a marker and pose message, given a file number to load the world_T_cam.